Owls 89504 - Common mounts MII sample generator changes, image creation scripts and draft documentation

ankedia opened the pull request:
1. Updated the MII sample generator scripts for creating the domain resources, building the images, and running the integration tests using the common mounts creating a WLS domain.
2. The changes include generating a build-image.sh script that can be used to create the common mounts image that hosts WDT model, archive, and WDT installation files.
3. Created initial draft of the common mounts documentation including steps for running the Model in Image initial use case using common mounts.

tbarnes-us commented on kubernetes/samples/scripts/create-weblogic-domain/model-in-image/cm-docker-file/Dockerfile:
```diff
@@ -0,0 +1,16 @@
+ARG GROUP=root
+ENV MOUNT_PATH=$MOUNT_PATH
+ENV WDT_MODEL_HOME=$MOUNT_PATH/models
+ENV WDT_HOME=${MOUNT_PATH}/weblogic-deploy
```
For clarity, can you make WDT_MODEL_HOME and WDT_HOME injected "ARG" values that are externally set to exactly match the values specified in the domain resource (instead of calculated values based off of MOUNT_PATH)? 

Also, rename WDT_HOME to WDT_INSTALL_HOME...  

Also please rename MOUNT_PATH to COMMON_MOUNT_PATH.  

Finally, add a comment to each of the ARG values that explains them?  For example COMMON_MOUNT_PATH/WDT_MODEL_HOME/WDT_INSTALL_HOME should each have a comment that mentions their domain resource equivalent...

tbarnes-us commented on operator/integration-tests/model-in-image/README:
```diff
@@ -41,6 +41,7 @@ then keep in mind that these are generated files. To modify them:
 # (1) modify the following as needed:
 #  ./mii-sample-wrapper/mii-domain.yaml.template-JRF
 #  ./mii-sample-wrapper/mii-domain.yaml.template-WLS
+#  ./mii-sample-wrapper/mii-domain.yaml.template-WLS-CM
```
Please add info about running in CM mode to line 33, something like:
```
Common Mount Mode - all 5 tests (initial, update1, update2, update3, update4), JRF optional:     
   ./run-test.sh -precleanup -oper -traefik -precleandb -cm [-jrf] -db -rcu -all
```

tbarnes-us commented on operator/integration-tests/model-in-image/mii-sample-wrapper/build-model-image.sh:
```diff
@@ -74,8 +74,37 @@ MODEL_ARCHIVE_FILES=$WORKDIR/$MODEL_DIR/archive.zip
+dryrun:
+dryrun:set -eux
+dryrun:
+dryrun:rm -rf $WORKDIR/cm-image/${MODEL_IMAGE_TAG}
```
This rm -rf is a bit scary IMO - the entire file system would be deleted if WORKDIR or MODEL_IMAGE_TAG were accidentally defined as `/ /tmp`. To at least partly defend, something like the following should hopefully limit the damage regardless of their value:
```
cd "$WORKDIR"
[ -d "cm-image/${MODEL_IMAGE_TAG}" ] && rm -rf "cm-image/${MODEL_IMAGE_TAG}"
```

tbarnes-us commented on operator/integration-tests/model-in-image/mii-sample-wrapper/build-model-image.sh:
```diff
@@ -74,8 +74,37 @@ MODEL_ARCHIVE_FILES=$WORKDIR/$MODEL_DIR/archive.zip
+if [[ ${MODEL_IMAGE_TAG} == *"-CM"* ]]; then
+cat << EOF
+dryrun:#!/bin/bash
+dryrun:# Use this script to build image '$MODEL_IMAGE_NAME:$MODEL_IMAGE_TAG'
```
to build image -> to build common mount image

tbarnes-us commented on operator/integration-tests/model-in-image/mii-sample-wrapper/build-model-image.sh:
```diff
@@ -74,8 +74,37 @@ MODEL_ARCHIVE_FILES=$WORKDIR/$MODEL_DIR/archive.zip
+dryrun:cp archive.zip $WORKDIR/cm-image/${MODEL_IMAGE_TAG}/models
+dryrun:cp archive.zip  $WORKDIR/$MODEL_DIR
+dryrun:cd $WORKDIR/cm-image/${MODEL_IMAGE_TAG}
+dryrun:#  Use additional docker build-arg as necessary to override default --build-arg=NAME=VALUE
```
To avoid repeat doc and allow for more detail, maybe replace these comments with something like `# see file $WORKDIR/cm-image/${MODEL_IMAGE_TAG} for an explanation of each --build-arg`.

tbarnes-us commented on operator/integration-tests/model-in-image/mii-sample-wrapper/README:
```diff
@@ -129,3 +129,11 @@ INTROSPECTOR_DEADLINE_SECONDS ::: Introspector timeout for JRF runs
 DB_NAMESPACE ::: Namespace for Oracle DB
 
 ?DB_NODEPORT ::: Nodeport for Oracle DB
+
```
Add IMAGE_TYPE explanation...

tbarnes-us replied:
And change MODEL_IMAGE_TAG explanation above to reflect its new dependency on IMAGE_TYPE

tbarnes-us commented on documentation/staging/content/userguide/managing-domains/model-in-image/common-mounts.md:
````diff
@@ -0,0 +1,444 @@
+    $ cd /tmp/mii-sample/archives/archive-v1
+    ```
+    ```
+    # Zip the archive to the location will later use when we run the WebLogic Image Tool
````
later ... --> later use when we use docker to build the common mount image

tbarnes-us commented on documentation/staging/content/userguide/managing-domains/model-in-image/common-mounts.md:
```diff
@@ -0,0 +1,444 @@
+  @@ [2020-04-30T13:50:42][seconds=0] Info:   image='model-in-image:WLS-v1'
+  @@ [2020-04-30T13:50:42][seconds=0] Info:   domainRestartVersion='1'
+  @@ [2020-04-30T13:50:42][seconds=0] Info:   namespace='sample-domain1-ns'
+  @@ [2020-04-30T13:50:42][seconds=0] Info:   domainUID='sample-domain1'
```
This is incorrect. It's a cut-and-paste from the old documentation for the non-common-mount case.

There are several things going on here that look like they need some investigation:

1) The current version of wl-pod-wait.sh 'image' is expected to be the base image since the script obtains the image name from the resource, and the resource image is the base image in the common mount use case.   So the 'image='model-in-image:WLS-v1'' output from wl-pod-wait.sh does not look it was run with the common mount use case.

2) On a different, but related note, the wl-pod-wait.sh script should ideally be updated to detect when the common mount image loaded by each init container matches the common mount image specified under 'serverPod' stanza for each pod.  I don't know if this information is even readily available on each pod, and even if it is available, I don't know if it'd be too complex to update the script to do this.  This is a problem, because if the common mount image is changed without also changing the domainRestartVersion, then the wl-pod-wait script can incorrectly assume that everything is up to date when it really isn't...

tbarnes-us replied:
1 has been addressed, 2 is still pending (am looking into it)

tbarnes-us commented on documentation/staging/content/userguide/managing-domains/model-in-image/common-mounts.md:
```diff
@@ -0,0 +1,444 @@
+{{% /notice %}}
+
+
+#### Common Mounts Fields
```
Propose consolidating this section and the next into a single section.  Something like:

Common Mount Reference

Use the `kubectl explain domain.spec.commonMountVolumes` and 
`kubectl explain domain.spec.serverPod.commonMounts` commands, or 
see the Domain Resource
[schema](https://github.com/oracle/weblogic-kubernetes-operator/blob/main/documentation/domains/Domain.md)
and [documentation]({{< relref "/userguide/managing-domains/domain-resource.md" >}}).

tbarnes-us commented on documentation/staging/content/userguide/managing-domains/model-in-image/common-mounts.md:
```diff
@@ -0,0 +1,444 @@
+### Running Model in Image initial use case using common mounts
+The initial use case for the Model in Image is described [here](/weblogic-kubernetes-operator/samples/simple/domains/model-in-image/initial/). The goal for this section is to create the initial domain using the common mounts feature where you provide the WDT model files and archive ZIP file in a separate container image.
+
+You will begin by following the steps described in the [initial use case sample](/weblogic-kubernetes-operator/samples/simple/domains/model-in-image/initial/). Once you have completed the entire use case, you will have the initial use case resources deployed and a running domain. To run the initial use case with common mounts, delete the existing domain by running the `kubectl delete domain sampl ...
```
Please break this paragraph into discrete steps, titled '#### Prerequisites'

tbarnes-us commented on documentation/staging/content/userguide/managing-domains/model-in-image/common-mounts.md:
````diff
@@ -0,0 +1,444 @@
+  $ docker run -it --rm model-in-image:v1 ls -l /common/weblogic-deploy
+  ```
+
+#### Domain Resource
````
This title could be more descriptive:  'Prepare and Apply the Domain Resource' or some-such...

tbarnes-us commented on documentation/staging/content/userguide/managing-domains/model-in-image/common-mounts.md:
````diff
@@ -0,0 +1,444 @@
+  ```
+  {{% /expand %}}
+
+You can compare this domain resource YAML file with the domain resource YAML file from the initial use case (`/tmp/mii-sample/domain-resources/WLS/mii-initial-d1-WLS-v1.yaml`) to see the changes required for the common mounts option. If you created your own YAML file, then you can make the required changes for the common mounts option and create the new domain using the modified YAML file.
````
I think it'd be helpful if right after the first sentence we added 'For example:', and then provided the diff command and its output right in a code block.

tbarnes-us commented on documentation/staging/content/userguide/managing-domains/model-in-image/common-mounts.md:
```diff
@@ -0,0 +1,444 @@
+When using the common mounts feature with the Model In Image use case, the WebLogic image would be the Domain resource's `domain.spec.image` field. The model files, application archives, and the WDT installation files can be provided in a separate image specified as part of the common mounts. The application archive could also be provided by a second container image in the common mounts section.  ...
+  - Use a WebLogic install image, or patch a WebLogic install image, without needing to include model artifacts within the image.
+  - Share one large WebLogic install image with multiple different model configurations.
+  - Distribute model files, application archives, and WebLogic Deploy Tooling executable using very small images instead of a large image that also contains a WebLogic install.
```
Proposed rewording:

Common mounts are an alternative approach for including Model in Image model files, application archive files, Weblogic Deploying Tooling install files, or other types of files, in your pods. This feature eliminates the need to provide these files in the image specified in `domain.spec.image`.  Instead:
- The domain resource's `domain.spec.image` directly reference a base image that only needs to include a WebLogic install and a Java install
- The domain resource's common mount related fields reference one or more smaller images that contain the desired Model in Image files.
- The domain resource `domain.spec.configuration.model.wdtInstallHome` and `domain.spec.configuration.model.modelHome` fields ares set to reference a directory that contains the files from the smaller images.

The common mounts feature internally uses a Kubernetes [emptyDir](https://kubernetes.io/docs/concepts/storage/volumes/#emptydir) volume and Kubernetes [init containers](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/) to share files from additional images within a WebLogic Server pod or the Introspector job pod.

The advantages of the common mounts feature for Model In Image domains are:  
- Use or patch a WebLogic install image without needing to include WDT install, application archive, or model artifacts within the image.  
- Share one large WebLogic install image with multiple different model configurations that are supplied in smaller images. 
- Distribute or update model files, application archives, and the WebLogic Deploy Tooling executable using very small images instead of a large image that also contains a WebLogic install.

tbarnes-us commented on documentation/staging/content/userguide/managing-domains/model-in-image/common-mounts.md:
```diff
@@ -0,0 +1,444 @@
+ - [Configuration](#configuration)
+ - [Common Mounts Fields](#common-mounts-fields)
+ - [Common Mount Volumes Fields](#common-mount-volumes-fields)
+ - [Running Model in Image initial use case using common mounts](#running-model-in-image-initial-use-case-using-common-mounts)
```
- I think it'd be more clear if the title was something like "Sample"

tbarnes-us commented on documentation/staging/content/userguide/managing-domains/model-in-image/common-mounts.md:
```diff
@@ -0,0 +1,444 @@
+  - Distribute model files, application archives, and WebLogic Deploy Tooling executable using very small images instead of a large image that also contains a WebLogic install.
+
+### Configuration
+Here's an example configuration for the Common Mounts. 
```
Accounting for feature gate:

Until this feature is officially released, the feature gate usage for the feature needs to mentioned in here and in the sample section below.  

It also needs to be integrated into the sample test: in the test script that deploys the operator when '-oper' is specified for 'run-test.sh' stand-alone (which in turn calls `deploy-operator.sh` which needs to modified to enable the feature gate), and in the java integration test itself when you update it to support common mounts.  (If the feature gate code is working, it should have been impossible for 'run-test.sh -oper -cm ...' stand-alone to pass without this modification.)

ankedia commented:
@tbarnes-us FYI - I have addressed some of the review comments above. I still need to look at the comments about the modifications to wl-pod-wait.sh script and accounting for feature gates. I'll look at those later and also go through comments one more time to make sure I haven't missed anything.

tbarnes-us commented on operator/integration-tests/model-in-image/mii-sample-wrapper/build-model-image.sh:
```diff
@@ -74,8 +74,32 @@ MODEL_ARCHIVE_FILES=$WORKDIR/$MODEL_DIR/archive.zip
+dryrun:cp archive.zip  $WORKDIR/$MODEL_DIR
+dryrun:cd $WORKDIR/cm-image/${MODEL_IMAGE_TAG}
+dryrun:# see file $WORKDIR/cm-image/${MODEL_IMAGE_TAG}/Dockerfile for an explanation of each --build-arg
+dryrun:docker build --build-arg COMMON_MOUNT_PATH=${COMMON_MOUNT_PATH} --build-arg WDT_MODEL_HOME=${WDT_MODEL_HOME} --build-arg WDT_INSTALL_HOME=${WDT_INSTALL_HOME} --tag ${MODEL_IMAGE_NAME}:${MODEL_IMAGE_TAG}  .
```
Issues:
1) not idempotent - the zip at line 94 will fail if it already occurred
2) docker build line too long for readability
3) it'd be more clear to break up into distinct steps using whitespace, cd to the working directory and use '.' throughout, and setup the archive zip using same pattern/location as the non-cm path

To address ^^^, suggest something like the following (not tested):
```
dryrun:set -eux
dryrun:
dryrun:rm -f $WORKDIR/$MODEL_DIR/archive.zip
dryrun:cd $WORKDIR/$ARCHIVE_SOURCEDIR
dryrun:zip -q -r $WORKDIR/$MODEL_DIR/archive.zip wlsdeploy
dryrun:
dryrun:cd "$WORKDIR"
dryrun:[ -d "cm-image/${MODEL_IMAGE_TAG}" ] && rm -rf cm-image/${MODEL_IMAGE_TAG}
dryrun:
dryrun:cd $WORKDIR/cm-image/${MODEL_IMAGE_TAG}
dryrun:cp $WORKDIR/$COMMON_MOUNT_DOCKER_FILE_SOURCEDIR/Dockerfile .
dryrun:mkdir ./models
dryrun:cp $MODEL_YAML_FILES ./models
dryrun:cp $MODEL_VARIABLE_FILES ./models
dryrun:cp $WORKDIR/$MODEL_DIR/archive.zip ./models
dryrun:unzip ${WORKDIR}/model-images/weblogic-deploy.zip -d .
dryrun:rm ./weblogic-deploy/bin/*.cmd
dryrun:
dryrun:# see file $WORKDIR/cm-image/${MODEL_IMAGE_TAG}/Dockerfile for an explanation of each --build-arg
dryrun:docker build --build-arg COMMON_MOUNT_PATH=${COMMON_MOUNT_PATH} \\
dryrun:             --build-arg WDT_MODEL_HOME=${WDT_MODEL_HOME} \\
dryrun:             --build-arg WDT_INSTALL_HOME=${WDT_INSTALL_HOME} \\
dryrun:             --tag ${MODEL_IMAGE_NAME}:${MODEL_IMAGE_TAG}  .
```

tbarnes-us commented on kubernetes/samples/scripts/create-weblogic-domain/model-in-image/cm-docker-file/Dockerfile:
```diff
@@ -0,0 +1,25 @@
+        COPY weblogic-deploy/ ${WDT_INSTALL_HOME}/
+        COPY models/ ${WDT_MODEL_HOME}/
+        RUN chown -R $USER:$GROUP ${COMMON_MOUNT_PATH}/
+USER $USER
```
Try avoid lines > 80 chars, maybe move the help text to top, plus add an explanation about the files purpose.  

For example:
```
...
# This is a sample Dockerfile for supplying Model in Image model files
# and a WDT install in a small separate "common mount" 
# image. This is an alternative to supplying the files directly
# in the domain resource `domain.spec.image` image.

# COMMON_MOUNT_PATH arg:
#   Parent location for Model in Image model and WDT installation files.
#   Must match domain resource 'domain.spec.commonMountVolumes.mountPath'
#   Default '/common'.
#
# WDT_MODEL_HOME arg:
#   Location for WebLogic Deploy Tooling model home.
#   Must match domain resource 'domain.spec.configuration.model.modelHome'
#   Must be a directory within COMMON_MOUNT_PATH.
#   Defaults to '${COMMON_MOUNT_PATH}/models'.
#
# WDT_INSTALL_HOME arg: 
#   Location of the WebLogic Deploy Tooling installation.
#   Must match domain resource 'domain.spec.configuration.model.wdtInstallHome'
#   Must be a directory within COMMON_MOUNT_PATH.
#   Defaults to '${COMMON_MOUNT_PATH}/weblogic-deploy'.

FROM busybox
ARG COMMON_MOUNT_PATH=/common
ARG WDT_INSTALL_HOME=${COMMON_MOUNT_PATH}/weblogic-deploy
ARG WDT_MODEL_HOME=${COMMON_MOUNT_PATH}/models
...
```

tbarnes-us commented on operator/integration-tests/model-in-image/mii-sample-wrapper/mii-domain.yaml.template-WLS-CM:
```diff
@@ -0,0 +1,154 @@
+        cpu: "250m"
+        memory: "768Mi"
+
+    # Settings for common mounts with images containing model, archives and WDT instal. See also 'spec.commonMountVolumes'.
```
Line too long. We could also supply more 'see also' here.  For example:
```
# Common mount image(s) containing WDT models, archives, and install. See also: 
#    'spec.commonMountVolumes'
#    'spec.configuration.model.modelHome'
#    'spec.configuration.model.wdtInstallHome'
```

tbarnes-us commented on operator/integration-tests/model-in-image/mii-sample-wrapper/mii-domain.yaml.template-WLS-CM:
```diff
@@ -0,0 +1,154 @@
+  # - "IF_NEEDED" will start all non-clustered servers, including the administration server, and clustered servers up to their replica count.
+  serverStartPolicy: "IF_NEEDED"
+
+  # settings for common mount volume, see also 'serverPod.commonMounts'.
```
```
# Settings for common mount volume(s), see also 'serverPod.commonMounts'.
```

tbarnes-us commented on operator/integration-tests/model-in-image/mii-sample-wrapper/generate-sample-doc.sh:
```diff
@@ -152,10 +152,11 @@ for phase in initial update1 update2 update3 update4; do
   $SCRIPTDIR/stage-and-create-ingresses.sh -dry
 
-  for type in WLS JRF; do
+  for IMAGE_TYPE in WLS WLS-CM JRF; do
```
To complete support for JRF in the CM case: add "JRF-CM" to the for loop, add a corresponding template, rerun this script (expect mismatch failure), check-in the newly generated sample files into the expected sample location, rerun (expect a pass this time).  The main test script should then pass for the '-jrf -cm' case.

tbarnes-us commented on operator/integration-tests/model-in-image/mii-sample-wrapper/env-init.sh:
```diff
@@ -19,11 +20,19 @@ if    [ ! "$WDT_DOMAIN_TYPE" = "WLS" ] \
+if    [ ! "$IMAGE_TYPE" = "WLS" ] \
+   && [ ! "$IMAGE_TYPE" = "WLS-CM" ] \
+   && [ ! "$IMAGE_TYPE" = "RestrictedJRF" ] \
+   && [ ! "$IMAGE_TYPE" = "JRF" ]; then
```
Add "JRF-CM"

tbarnes-us commented on operator/integration-tests/model-in-image/README:
```diff
@@ -52,6 +56,7 @@ rm -fr /tmp/mii-sample   # deletes default output location for the following com
 cp /tmp/mii-sample/ingresses/*.yaml ../../../../kubernetes/samples/scripts/create-weblogic-domain/model-in-image/ingresses
 cp /tmp/mii-sample/domain-resources/JRF/*.yaml ../../../../kubernetes/samples/scripts/create-weblogic-domain/model-in-image/domain-resources/JRF
```
Add cp for JRF-CM

tbarnes-us commented on operator/integration-tests/model-in-image/deploy-operator.sh:
```diff
@@ -65,6 +65,7 @@ helm install $OPER_NAME kubernetes/charts/weblogic-operator \
   --set       serviceAccount=$OPER_SA \
   --set       "domainNamespaces={$DOMAIN_NAMESPACE}" \
   --set       "javaLoggingLevel=INFO" \
+  --set       "featureGates=CommonMounts=${DO_CM}" \
```
This script is designed so that it can run independently of run-test.sh if desired (where DO_CM is defined).
Instead of expecting DO_CM to be set, it'd probably be fine just to hard code it to true.

ankedia commented:
@rosemarymarano If possible, please make your edits/changes directly in the PR and commit them. Thanks.

rosemarymarano reviewed (approved):
Incorporated my edits.

rjeberhard commented on documentation/staging/content/userguide/managing-domains/model-in-image/common-mounts.md:
```diff
@@ -0,0 +1,678 @@
+
+- Use or patch a WebLogic installation image without needing to include a WDT installation,
+  application archive, or model artifacts within the image.
+- Share one large WebLogic installation image with multiple different model
```
Let's not highlight that WebLogic images are particularly large ;). Perhaps remove the word "large" and "smaller" becomes "specific".

tbarnes-us replied:
Fixed.

rjeberhard commented on documentation/staging/content/userguide/managing-domains/model-in-image/common-mounts.md:
```diff
@@ -0,0 +1,678 @@
+- Share one large WebLogic installation image with multiple different model
+  configurations that are supplied in smaller images.
+- Distribute or update model files, application archives, and the
+  WebLogic Deploy Tooling executable using very small images,
```
Similar... What if we just remove the "instead" phrase.

tbarnes-us replied:
Fixed.

tbarnes-us replied:
Please also look at changes to documentation/staging/content/userguide/managing-domains/choosing-a-model/_index.md

rjeberhard commented on documentation/staging/content/userguide/managing-domains/model-in-image/common-mounts.md:
````diff
@@ -0,0 +1,678 @@
+  - For example, assuming that
+    you have set up `/tmp/mii-sample` as your working directory:
+    ```shell
+    kubectl delete domain sample-domain1 -n sample-domain1-ns
````
Add $ at beginning of shell command line

tbarnes-us replied:
Fixed.

rjeberhard commented on documentation/staging/content/userguide/managing-domains/model-in-image/common-mounts.md:
````diff
@@ -0,0 +1,678 @@
+
+1. Create a temporary directory for staging the common mount image's files and `cd` to this directory:
+   ```shell
+   mkdir /tmp/mii-sample/cm-image/WLS-CM-v1
````
$ here too... I won't mark them all, please please review.

tbarnes-us replied:
Fixed.

rjeberhard commented on documentation/staging/content/userguide/managing-domains/model-in-image/common-mounts.md:
````diff
@@ -0,0 +1,678 @@
+  ```
+  ```
+  NAME                                READY   STATUS    RESTARTS   AGE
+  sample-domain1-introspector-z5vmp   0/1     Pending   0          0s
````
I wonder why this content doesn't line up... Are there tabs or spaces here? Can you check how this looks from Hugo?

tbarnes-us replied:
It doesn't line up because that's the way the watch heuristic works - watch cannot predict if/when it'll need to widen columns as it continues to report...

rjeberhard commented on documentation/staging/content/userguide/managing-domains/model-in-image/overview.md:
```diff
@@ -24,36 +24,49 @@ Unlike Domain in PV and Domain in Image, Model in Image eliminates the need to p
- - Updating WDT model files at runtime. For example, you can add a data source to a running domain. See [Runtime updates](#runtime-updates) for details.
+ - Embedding these files in a single image that also contains a WebLogic installation,
+   and using the WebLogic Image Tool (WIT) to generate this image. Or, alternatively,
+   embedding the files in one or more small images.
```
More "small"... I'm still worried about perspective and would prefer "application-specific" or similar.

tbarnes-us replied:
Fixed.

rjeberhard commented on kubernetes/samples/scripts/create-weblogic-domain/model-in-image/cm-docker-file/Dockerfile:
```diff
@@ -0,0 +1,40 @@
+        RUN adduser -D -u ${USERID} -G $GROUP $USER
+        COPY weblogic-deploy/ ${WDT_INSTALL_HOME}/
+        COPY models/ ${WDT_MODEL_HOME}/
+        RUN chown -R $USER:$GROUP ${COMMON_MOUNT_PATH}/
```
You can do the chown as part of the COPY command rather than as a separate RUN... https://docs.docker.com/engine/reference/builder/#copy

rjeberhard replied:
It's probably more efficient to chown the top-level COMMON_MOUNT_PATH in the same RUN where you add the user and then chown on the two copies... Doing a chown after the copy doubles the size of those files in the image.

ankedia replied:
I tried `chown` on the copies using USER and GROUP variables (like below) but it fails with the following error -
COPY --chown=${USER}:${GROUP} weblogic-deploy/ ${WDT_INSTALL_HOME}/

```
Step 12/14 : COPY --chown=${USER}:${GROUP} weblogic-deploy/ ${WDT_INSTALL_HOME}/
unable to convert uid/gid chown string to host mapping: can't find uid for user ${USER}: no such user: ${USER}
```
This works if I hard code the user and group during COPY -
```
COPY --chown=oracle:root weblogic-deploy/ ${WDT_INSTALL_HOME}/
```
This issue is reported in https://github.com/moby/moby/issues/37290 (which is closed) but I still see it in docker version `19.03.11-ol, build 2a9c376`. For now, I left it as it is, please let me know if I missed something. Thanks.

rjeberhard commented on kubernetes/samples/scripts/create-weblogic-domain/model-in-image/cm-docker-file/Dockerfile:
```diff
@@ -0,0 +1,40 @@
+ENV COMMON_MOUNT_PATH=${COMMON_MOUNT_PATH}
+ENV WDT_MODEL_HOME=${WDT_MODEL_HOME}
+ENV WDT_INSTALL_HOME=${WDT_INSTALL_HOME}
+        RUN adduser -D -u ${USERID} -G $GROUP $USER
```
Why are these tabbed over?

rjeberhard reviewed (approved):
I have a couple of minor issues... please do take a look at those, but they are relatively minor, so I'm going to pre-approve now.

tbarnes-us reviewed (approved):
Looks good!

One remaining issue, but it's not a blocker:  I am in the midst of updating the 'wl-pod-wait' script to check whether pods have their expected common mount images (right now, it checks for the main image, restartVersion, introspectVersion, and ready state). This script is used by CAGBU, the sample, and integration testing...

tbarnes-us commented on kubernetes/samples/scripts/create-weblogic-domain/model-in-image/cm-docker-file/Dockerfile:
```diff
@@ -0,0 +1,40 @@
+ENV WDT_INSTALL_HOME=${WDT_INSTALL_HOME}
+RUN adduser -D -u ${USERID} -G $GROUP $USER
+COPY weblogic-deploy/ ${WDT_INSTALL_HOME}/
+COPY models/ ${WDT_MODEL_HOME}/
```
Minor: Ideally, the source directories for INSTALL_HOME and MODEL_HOME would be configurable (WDT_INSTALL_SOURCE? WDT_MODEL_SOURCE?).  Or, perhaps better, it might be more generic and easier to understand if the script removed WDT_INSTALL_HOME and WDT_MODEL_HOME and instead simply had COMMON_MOUNT_PATH and COMMON_MOUNT_SOURCE...

ankedia replied:
Done in 3d6c9bc and abf94f5. Let me know if I missed anything.

tbarnes-us replied:
Looks good!
